src/integrationtests/java/com/aws/greengrass/integrationtests/metrics/MetricsEmitterTest.java

```diff
@@ -0,0 +1,207 @@
+                new DeviceConfiguration(kernel, MOCK_THING_NAME, "us-east-1", "us-east-1", "mock", "mock", "mock",
+                        "us-east-1", "mock"));
+
+        //Emit metric events
+        domainEvents.emit(new CertificateSubscriptionEvent(GetCertificateRequestOptions.CertificateType.SERVER,
+                CertificateSubscriptionEvent.SubscriptionStatus.SUCCESS));
+        domainEvents.emit(new AuthorizeClientDeviceActionEvent(AuthorizeClientDeviceActionEvent
+                .AuthorizationStatus.SUCCESS));
+        domainEvents.emit(new VerifyClientDeviceIdentityEvent(VerifyClientDeviceIdentityEvent
+                .VerificationStatus.SUCCESS));
+        domainEvents.emit(new SessionCreationEvent(SessionCreationEvent.SessionCreationStatus.SUCCESS));
+        domainEvents.emit(new ServiceErrorEvent());
+
+        //WHEN
+        startNucleusWithConfig("metricsConfig.yaml");
+        int periodicAggregateIntervalSec = kernel.getContext().get(MetricsConfiguration.class).getAggregatePeriod();
+        assertEquals(AGGREGATE_INTERVAL, periodicAggregateIntervalSec);
+        assertNotNull(kernel.getContext().get(MetricsEmitter.class).getFuture(),
+                "periodic publish future is not scheduled");
+        long delay = kernel.getContext().get(MetricsEmitter.class).getFuture().getDelay(TimeUnit.SECONDS);
+        assertEquals(delay, periodicAggregateIntervalSec);
+
+        //move clock one day ahead to trigger publish
+        Instant aDayLater = Instant.now().plusSeconds(DEFAULT_PUBLISH_INTERVAL);
+        mockInstant(aDayLater.toEpochMilli());
+
+        //check for telemetry logs in ~root/telemetry
+        assertEquals(kernel.getNucleusPaths().rootPath().resolve("telemetry"),
+                TelemetryConfig.getTelemetryDirectory());
+
+        //THEN
+        boolean telemetryMessageVerified = false;
+        String telemetryPublishTopic = DEFAULT_TELEMETRY_METRICS_PUBLISH_TOPIC
+                .replace("{thingName}", MOCK_THING_NAME);
+        CountDownLatch metricsPublished = new CountDownLatch(1);
+        List<PublishRequest> requests = new ArrayList<>();
+        when(mqttClient.publish(any(PublishRequest.class))).thenAnswer(i -> {
+            Object argument = i.getArgument(0);
+            PublishRequest publishRequest = (PublishRequest) argument;
+            if (telemetryPublishTopic.equals(publishRequest.getTopic())) {
+                requests.add(publishRequest);
+                metricsPublished.countDown();
+            }
+            return CompletableFuture.completedFuture(0);
+        });
+        assertTrue(metricsPublished.await(30, TimeUnit.SECONDS), "Metrics not published");
+
+        for (PublishRequest request : requests) {
+            if (!telemetryPublishTopic.equals(request.getTopic())) {
+                continue;
+            }
+            try {
+                MetricsPayload mp = new ObjectMapper().readValue(request.getPayload(), MetricsPayload.class);
```

Comment:
you need to validate the actual metrics. It is good that you're validating _something_ was published, but you need to make sure it contains all the metrics which you tried to push (and that they're aggregated).

Reply:
Just to double check, will the mqtt publish message have the same structure as the messages in the telemetry log in here: [https://github.com/aws-greengrass/aws-greengrass-nucleus/tree/main/src/main/java/com/aws/greengrass/telemetry](url)

Reply:
https://github.com/aws-greengrass/aws-greengrass-nucleus/blob/main/src/main/java/com/aws/greengrass/telemetry/TelemetryAgent.java#L262. It pushes using the metric payload type.